// _heirloom-core@2.0.6@heirloom-core/lib/engines/KoaEngine.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _koa = require('koa');

var _koa2 = _interopRequireDefault(_koa);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const pkg = require('koa/package.json'); /**
                                          * 
                                          * @author xuyuanxiang
                                          * @date 2017/2/3
                                          */
class KoaEngine {

    static shareInstance() {
        if (!this.defaultInstance) {
            this.defaultInstance = new KoaEngine();
        }
        return this.defaultInstance;
    }

    constructor() {
        this.name = pkg.name;
        this.version = pkg.version;

        this.koa = new _koa2.default();
    }

    use(...generator) {
        generator.forEach(it => this.koa.use(it));
    }

    listen(port, callback) {
        this.koa.listen(port, callback);
    }

    on(event, callback) {
        this.koa.on(event, callback);
    }
}
exports.default = KoaEngine;
